src/foam/dao/QueryCachingDAODecorator.js

```diff
@@ -0,0 +1,156 @@
+        self.refreshDaoCount_(self).then(function() {
+
+          var requestStartIdx = typeof skip !== 'undefined' ? skip : 0;
+          var requestEndIdx = typeof limit !== 'undefined' && skip + limit < self.daoCount_ ? skip + limit : self.daoCount_;
+
+          // Ensure we have cache for request
+          self.fillCache_(self, key, requestStartIdx, requestEndIdx, x, sink, order, predicate).then(function() {
+
+            // Return data from cache
+            for (let idx = requestStartIdx; idx < requestEndIdx; idx++) {
+              sink.put(self.cache[key][idx]);
+            }
+
+            resolve(sink);
+          });
+        });
+      });
+    },
+
+    // Remove invalidates cache and is forwarded to the source.
+    function remove_(x, o) {
+      this.cache = {};
+      this.clearProperty('daoCount_');
+      return this.delegate.remove_(x, o);
+    },
+
+    // RemoveAll invalidates cache and is forwarded to the source.
+    function removeAll_(x, skip, limit, order, predicate) {
+      this.cache = {};
+      this.clearProperty('daoCount_');
+      this.delegate.removeAll_(x, skip, limit, order, predicate);
+    },
+
+    function cmd_(x, obj) {
+      if ( obj == this.CLEAR_CACHE ) {
+        this.cache = {};
+        this.clearProperty('daoCount_');
+      } else {
+        this.SUPER(x, obj);
+      }
+    },
+
+    function fillCache_(self, key, requestStartIdx, requestEndIdx, x, sink, order, predicate) {
+      // Pre-check cache for elements
+      let startIdx = -1;
+      let endIdx = -1;
+      let hasMissingData = false;
+
+      if (self.cache[key]) {
+        // Cycle through exising cached elements to verify all requested are present
+        for ( let idx = requestStartIdx; idx < requestEndIdx; idx++ ) {
+          if ( ! self.cache[key][idx] ) {
+            if ( ! hasMissingData ) {
+              // Found start of missing data withing requested block
+              hasMissingData = true;
+              startIdx = idx;
+            }
+
+            // Make sure we have the last index of a missing data element within requested block
+            endIdx = idx + 1;
+          }
+        }
+      } else {
+        // No data present load entire block
+        hasMissingData = true;
+        startIdx = requestStartIdx;
+        endIdx = requestEndIdx;
+        self.cache[key] = [];
+      }
+
+      if (hasMissingData) {
+        console.log('******** QUERYCACHE*** HAS MISSING DATA ***: key: ' + key + 'daoCount: ' + self.daoCount_ + ' startIdx: ' + startIdx + ' endIdx: ' + endIdx);
+        return this.delegate.select_(x, sink, startIdx, 1 + endIdx - startIdx, order, predicate).then(function (result) {
+          // Point to the appropriate array source
+//          let array = foam.mlang.sink.Projection.isInstance(sink) ? result.projectionWithClass : result.array;
+
+          // Update cache with missing data
+          for (let idx = 0; idx < result.array.length; idx++) {
```

Comment:
Spacing: for ( let idx = 0 ; idx < result.array.length ; idx++ ) {

Reply:
Updated